[SPARK-48320][CORE][DOCS] Add structured logging guide to the scala and java doc

panbingkun opened the pull request:
### What changes were proposed in this pull request?
The pr aims to add `external third-party ecosystem access` guide to the `scala/java` doc.

The external third-party ecosystem is very extensive. Currently, the document covers two scenarios:
- Pure java (for example, an application only uses the java language - many of our internal production applications are like this)
- java + scala

### Why are the changes needed?
Provide instructions for external third-party ecosystem access to the structured log framework.


### Does this PR introduce _any_ user-facing change?
Yes, When an external third-party ecosystem wants to access the structured log framework, developers can get help through this document.


### How was this patch tested?
- Add new UT.
- Manually test.
- Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

panbingkun committed:
[SPARK-48320][CORE][DOCS] Add external third-party ecosystem access guide to the doc

panbingkun commented on common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:
```diff
@@ -22,7 +22,7 @@ import java.util.Locale
  */
 trait LogKey {
-  val name: String = this.toString.toLowerCase(Locale.ROOT)
+  def name: String = this.getClass.getSimpleName.stripSuffix("$").toLowerCase(Locale.ROOT)
```
If we don't make such changes, the `JavaCustomLogKeys`(Custom `LogKey` that may be extended by `external third-party ecosystems`) class will have to be defined like this, which is very ugly and looks weird.
<img width="772" alt="image" src="https://github.com/apache/spark/assets/15246973/9e621224-254c-4b62-8aaa-2973dd4e7b3a">

panbingkun commented on common/utils/src/test/java/org/apache/spark/util/SparkLoggerSuiteBase.java:
```diff
@@ -104,8 +112,8 @@ public void testBasicMsgLogger() {
       try {
-        assert (captureLogOutput(pair.getRight()).matches(
-            expectedPatternForBasicMsg(pair.getLeft())));
+        Assertions.assertTrue(captureLogOutput(pair.getRight()).matches(
```
replace `assert` with `Assertions.assertTrue`

panbingkun commented on common/utils/src/test/java/org/apache/spark/util/SparkLoggerSuiteBase.java:
```diff
@@ -140,15 +148,13 @@ public void testLoggerWithMDC() {
     Runnable errorFn = () -> logger().error(msgWithMDC, executorIDMDC);
     Runnable warnFn = () -> logger().warn(msgWithMDC, executorIDMDC);
     Runnable infoFn = () -> logger().info(msgWithMDC, executorIDMDC);
-    Runnable debugFn = () -> logger().debug(msgWithMDC, executorIDMDC);
```
After https://github.com/apache/spark/pull/46405, the following 2 lines of code are no longer needed.

panbingkun commented:
cc @gengliangwang

gengliangwang commented on common/utils/src/main/scala/org/apache/spark/internal/LogKey.scala:
```diff
@@ -22,7 +22,7 @@ import java.util.Locale
  */
 trait LogKey {
-  val name: String = this.toString.toLowerCase(Locale.ROOT)
+  def name: String = this.getClass.getSimpleName.stripSuffix("$").toLowerCase(Locale.ROOT)
```
shall we have a val or a lazy val to memorize the key?

panbingkun replied:
Okay, let's use `lazy val`, I have tested it and it works well.

panbingkun replied:
<img width="789" alt="image" src="https://github.com/apache/spark/assets/15246973/e9d5ee39-c1cd-431a-ac9a-648a684ddf9f">

gengliangwang commented on common/utils/src/main/scala/org/apache/spark/internal/README.md:
````diff
@@ -45,3 +45,52 @@ logger.error("Failed to abort the writer after failing to write map output.", e)
+  public static final CUSTOM_LOG_KEY CUSTOM_LOG_KEY = new CUSTOM_LOG_KEY();
+}
+```
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;and use it in `java` code:
````
why do we have &nbsp here?

panbingkun replied:
If `&nbsp` is not added, its `indent` looks incorrect, as shown below:
<img width="1131" alt="image" src="https://github.com/apache/spark/assets/15246973/18474be8-ada9-456e-bc08-decb46b74e3c">

Let me incorporate this sentence into the above one, and revise it as follows:
<img width="1133" alt="image" src="https://github.com/apache/spark/assets/15246973/45e289ad-e948-4f4f-8789-173c156b7c44">

gengliangwang commented on common/utils/src/main/scala/org/apache/spark/internal/README.md:
````diff
@@ -45,3 +45,52 @@ logger.error("Failed to abort the writer after failing to write map output.", e)
+* If you want to output logs in `scala code` through the structured log framework, you can define `custom LogKeys` as follows:
+
+```scala
+object CustomLogKeys {
````
let's not mention CustomLogKeys here. It is not necessary, right?

panbingkun replied:
Okay

gengliangwang commented on common/utils/src/main/scala/org/apache/spark/internal/README.md:
````diff
@@ -45,3 +45,52 @@ logger.error("Failed to abort the writer after failing to write map output.", e)
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;when your project has **depended on `scala`**,
+```scala
+object CustomLogKeys {
````
the content seems duplicated?

panbingkun replied:
<img width="1112" alt="image" src="https://github.com/apache/spark/assets/15246973/dae45ab2-9c49-40aa-a296-ebb30d148142">

1.Use the custom LogKey defined in the `scala code`, and use it in the `scala code`
2.Use the custom LogKey defined in the `java code`, and use it in the `java code`
3.Use the custom LogKey defined in the `scala code`, and use it  in the `java code`

panbingkun replied:
Do we need to delete the description of the last case - 3?

panbingkun commented:
<img width="1127" alt="image" src="https://github.com/apache/spark/assets/15246973/b0e4a64e-5144-4f6e-9e34-151758f19746">

panbingkun committed:
update

panbingkun committed:
Merge branch 'master' into SPARK-48320

panbingkun committed:
fix

panbingkun committed:
Merge branch 'master' into SPARK-48320

mridulm commented on common/utils/src/main/scala/org/apache/spark/internal/README.md:
````diff
@@ -45,3 +45,29 @@ logger.error("Failed to abort the writer after failing to write map output.", e)
+import org.apache.spark.internal.MDC;
+
+logger.error("Unable to delete key {} for cache", MDC.of(CUSTOM_LOG_KEY, "key"));
+```
````
Spark logging is not exposed api for users to directly leverage.

panbingkun replied:
If a third-party ecosystem wants to access a structured log framework, should it customize one?

mridulm replied:
I am not sure I follow the question.

Is spark Logging being used today by third party ? Yes.
Do we support it ? No.
Do we disallow it ? Not really - power users put all sorts of things in org.apache.spark namespace and access internals, but we don't support that.

We are able to make these changes for structured logging, potentially breaking some usecases out there (even if we are making efforts to minimize impact) because this is an internal api.

panbingkun replied:
1.Let me take an example. In the production system, for example, an application named `SparkPi`, if you want to use `log4j` to print logs, there are two ways,
One is as follows(It does not need to be placed under the `org.apache.spark namespace`. `Logging` can also be used):
<img width="519" alt="image" src="https://github.com/apache/spark/assets/15246973/2f0c1cca-122c-4d3d-b67d-f64eae3f0ca1">

The other is:
<img width="497" alt="image" src="https://github.com/apache/spark/assets/15246973/04ba4b07-2442-4630-8aea-4e6e632bcaf4">

The last one is often used by `me`, and this is just `an application`.

If a third-party system, such as `iceberg`, wants to summarize logs through `a structured log framework`, it should be easy to do so by using the current structured framework.

2.In fact, our logging does not indicate that it is `any internal API` (We just didn't mark it as Developer Api), as shown below:
https://github.com/apache/spark/blob/fa8aa571ad18441622bb7e3ac66032ab9e7cbc0a/common/utils/src/main/scala/org/apache/spark/internal/Logging.scala#L84

I am worried that many users have used it like me, because it is direct and convenient (`conf/log4j.properties` can be used together)

mridulm replied:
Like I said, are people using it ? Yes.
Is it exposed api, no.

The logging code is within 'internal' package.

gengliangwang replied:
@mridulm the modified file here is `common/utils/src/main/scala/org/apache/spark/internal/README.md`, a README under the internal package.  It is not part of the Spark doc site.

I don't have a strong opinion on this though. If you insist we should NOT put this into the doc here, we should enhance the scala/java doc instead.

mridulm replied:
Let us drop this doc - this feature is essentially private to spark. Let us avoid documenting it such that it gives an impression that users can directly leverage it.

panbingkun replied:
Well, I will remove the above `README.md` file and write the above guide to the `scala/java` doc.

panbingkun replied:
@mridulm @gengliangwang 
The `common/utils/src/main/scala/org/apache/spark/internal/README.md` file has been dropped, and the content of the original document has been distributed to `java/scala` docs.

panbingkun committed:
Merge branch 'master' into SPARK-48320

panbingkun committed:
remove README.md

panbingkun committed:
update

mridulm commented on common/utils/src/main/java/org/apache/spark/internal/SparkLogger.java:
```diff
@@ -28,6 +28,50 @@
+ *   you can define `custom LogKey` and use it in `java` code as follows:
+ * <p>
+ *
+ * // External third-party ecosystem `custom LogKey` must be `implements LogKey`
```
```suggestion
 * // To add a `custom LogKey`, implement `LogKey`
```

mridulm commented on common/utils/src/main/java/org/apache/spark/internal/SparkLogger.java:
```diff
@@ -28,6 +28,50 @@
+ * logger.error("Failed to abort the writer after failing to write map output.", e);
+ * <p>
+ *
+ * External third-party ecosystem access:
```
```suggestion
```

mridulm commented on common/utils/src/main/scala/org/apache/spark/internal/Logging.scala:
```diff
@@ -29,6 +29,45 @@ import org.slf4j.{Logger, LoggerFactory}
+ *   you can define `custom LogKey` and use it in `scala` code as follows:
+ * <p>
+ *
+ * // External third-party ecosystem `custom LogKey` must be `extends LogKey`
```
```suggestion
 * // To add a `custom LogKey`, implement `LogKey`
```

mridulm commented on common/utils/src/test/java/org/apache/spark/util/PatternSparkLoggerSuite.java:
```diff
@@ -87,4 +87,10 @@ String expectedPatternForMsgWithMDCValueIsNull(Level level) {
   }
+
+  @Override
+  String expectedPatternForExternalSystemJavaCustomLogKey(Level level) {
```
```suggestion
  String expectedPatternForJavaCustomLogKey(Level level) {
```

mridulm commented on common/utils/src/test/java/org/apache/spark/util/PatternSparkLoggerSuite.java:
```diff
@@ -87,4 +87,10 @@ String expectedPatternForMsgWithMDCValueIsNull(Level level) {
+  @Override
+  String expectedPatternForExternalSystemJavaCustomLogKey(Level level) {
+    return toRegexPattern(level,
+ ".*<level> <className>: External system custom log message.\n");
```
```suggestion
 ".*<level> <className>: custom log message.\n");
```

mridulm commented on common/utils/src/test/java/org/apache/spark/util/SparkLoggerSuiteBase.java:
```diff
@@ -69,6 +70,9 @@ private String basicMsg() {
   private final MDC externalSystemCustomLog =
     MDC.of(CustomLogKeys.CUSTOM_LOG_KEY$.MODULE$, "External system custom log message.");
 
+  private final MDC externalSystemJavaCustomLog =
+    MDC.of(JavaCustomLogKeys.CUSTOM_LOG_KEY, "External system custom log message.");
```
```suggestion
  private final MDC javaCustomLogMdc =
    MDC.of(JavaCustomLogKeys.CUSTOM_LOG_KEY, "Custom log message.");
```

mridulm commented on common/utils/src/test/java/org/apache/spark/util/SparkLoggerSuiteBase.java:
```diff
@@ -90,6 +94,9 @@ private String basicMsg() {
   abstract String expectedPatternForExternalSystemCustomLogKey(Level level);
 
+  // test for external system java custom LogKey
+  abstract String expectedPatternForExternalSystemJavaCustomLogKey(Level level);
```
```suggestion
  abstract String expectedPatternForJavaCustomLogKey(Level level);
```

mridulm commented on common/utils/src/test/java/org/apache/spark/util/SparkLoggerSuiteBase.java:
```diff
@@ -227,4 +232,30 @@ public void testLoggerWithExternalSystemCustomLogKey() {
   }
+
+  @Test
+  public void testLoggerWithExternalSystemJavaCustomLogKey() {
```
```suggestion
  public void testLoggerWithJavaCustomLogKey() {
```

mridulm commented on common/utils/src/test/java/org/apache/spark/util/StructuredSparkLoggerSuite.java:
```diff
@@ -161,5 +161,19 @@ String expectedPatternForExternalSystemCustomLogKey(Level level) {
   }
+
+  @Override
+  String expectedPatternForExternalSystemJavaCustomLogKey(Level level) {
```
```suggestion
  String expectedPatternForJavaCustomLogKey(Level level) {
```

mridulm commented on common/utils/src/test/java/org/apache/spark/util/PatternSparkLoggerSuite.java:
```diff
@@ -87,4 +87,10 @@ String expectedPatternForMsgWithMDCValueIsNull(Level level) {
   String expectedPatternForExternalSystemCustomLogKey(Level level) {
     return toRegexPattern(level, ".*<level> <className>: External system custom log message.\n");
   }
```
```suggestion
  String expectedPatternForCustomLogKey(Level level) {
    return toRegexPattern(level, ".*<level> <className>: custom log message.\n");
  }
```

mridulm commented on common/utils/src/test/java/org/apache/spark/util/SparkLoggerSuiteBase.java:
```diff
@@ -69,6 +70,9 @@ private String basicMsg() {
   private final MDC externalSystemCustomLog =
     MDC.of(CustomLogKeys.CUSTOM_LOG_KEY$.MODULE$, "External system custom log message.");
 
```
```suggestion
  private final MDC customLog =
    MDC.of(CustomLogKeys.CUSTOM_LOG_KEY$.MODULE$, "custom log message.");

```

mridulm reviewed (changes requested):
Please replace all references to `external` and `external system` - we dont want to set expectations that external system integrations are encouraged.
Keep it as custom integrations instead - if power users and frameworks want to leverage the integration, this would be the path for them as it is not explicitly `private [spark]`.

I added a few cases below, but I would expect @gengliangwang to be better aware and give much better guidance than what I can of what else might need to be changed - as he is more actively reviewing the PR's :-)

panbingkun committed:
Apply suggestions from code review

Co-authored-by: Mridul Muralidharan <1591700+mridulm@users.noreply.github.com>

panbingkun committed:
update

panbingkun committed:
Merge branch 'master' into SPARK-48320

panbingkun committed:
update

panbingkun commented:
> Please replace all references to `external` and `external system` - we dont want to set expectations that external system integrations are encouraged. Keep it as custom integrations instead - if power users and frameworks want to leverage the integration, this would be the path for them as it is not explicitly `private [spark]`.
> 
> I added a few cases below, but I would expect @gengliangwang to be better aware and give much better guidance than what I can of what else might need to be changed - as he is more actively reviewing the PR's :-)

Thank you very much for your patient and careful review. ❤️
I have updated it again.

mridulm reviewed (approved):
Thanks for the changes @panbingkun.
Looks good to me - I will let @gengliangwang review and merge; he is much more knowledgeable about this in general !

gengliangwang commented on common/utils/src/main/scala/org/apache/spark/internal/README.md:
@mridulm @panbingkun This README is for Spark developers. As long as it doesn't mention 3rd party supports, can we just keep it?

panbingkun replied:
I'm fine to it, WDYT @mridulm ?
we can just put the new description in the `scala/java` doc.

mridulm replied:
IIRC @panbingkun included the contents into the existing code; which is also a nice addition to the code documentation.
IMO that is a better overall approach given this is internal to spark.

panbingkun replied:
Okay, I have adopted the plan of deleting `README.md`

gengliangwang commented:
LGTM except for one comment.

panbingkun committed:
Merge branch 'master' into SPARK-48320

panbingkun committed:
restore README.md

panbingkun committed:
Merge branch 'master' into SPARK-48320

panbingkun committed:
Revert "restore README.md"

This reverts commit a5fcf8f4c07dfa068a912c39192f581669c2bdcb.

gengliangwang commented:
Thanks, merging to master
